Add support for EFFECT_OFF

dmulcahey opened the pull request:
Fixes: #98

dmulcahey committed:
Add support for EFFECT_NONE

dmulcahey commented:
@puddly @TheJulianJES do we want to do this color mode part: https://developers.home-assistant.io/blog/2024/01/23/light-color-mode-changes/#requirements-on-color_mode-are-less-strict-when-a-light-is-rendering-an-effect

TheJulianJES commented:
Can't comment on unchanged lines, but the assume option for effects should be changed to be like the others.
So, adding the `is not None` check. Line 1383 and following currently.
![image](https://github.com/user-attachments/assets/71ecdeeb-18aa-4b4d-ac24-d2cdd029f355)

---

> do we want to do this color mode part

We could do that, yeah. (would prefer a separate PR though)
IIRC, changing the color during a colorloop effect doesn't impact the bulb, so we could set the color modes to on_off and brightness during that. Brightness changes still affect it, I think.

(On a semi-related note, I'd also really like to support Hue effects in the future. Will require changes to both quirks and zha.)

TheJulianJES commented on zha/application/platforms/light/const.py:
```diff
@@ -91,6 +91,7 @@ class ColorMode(StrEnum):
 EFFECT_COLORLOOP: Final[str] = "colorloop"
 EFFECT_RANDOM: Final[str] = "random"
 EFFECT_WHITE: Final[str] = "white"
+EFFECT_NONE: Final[str] = "None"
```
Shouldn't this be `EFFECT_OFF = "off"` (not `EFFECT_NONE`)?
At least, that's what HA uses: [home-assistant/core/blob/homeassistant/components/light/__init__.py#L228-L233](https://github.com/home-assistant/core/blob/7820bcf2186c9e9d815c91c724e5187809b64d4b/homeassistant/components/light/__init__.py#L228-L233)

dmulcahey replied:
Yep not sure wth I was looking at 😂. Good catch

dmulcahey committed:
off not None

dmulcahey committed:
review comment

TheJulianJES commented on zha/application/platforms/light/__init__.py:
```diff
@@ -1044,8 +1045,8 @@ def _assume_group_state(self, update_params) -> None:
             # the effect is always deactivated in async_turn_on if not provided
-            if effect is None:
-                self._effect = None
+            if effect is None or effect == EFFECT_OFF:
+                self._effect = EFFECT_OFF
             elif self._effect_list and effect in self._effect_list:
                 self._effect = effect
```
I believe this still needs some changes, though I haven't tested the PR yet.
This might be what we actually want here:
```suggestion
            if effect is not None and effect in self._effect_list:
                self._effect = effect
```
(It's `None` if the `effect` wasn't touched on the group. In that case, we do not want to modify the member lights. If it's not None and in the effect list, the member will have applied that effect, so we want to set that state too. This will include `EFFECT_OFF`.)

EDIT: `_effect_list` can be `None` though, so maybe:
```suggestion
            if effect is not None and self._effect_list and effect in self._effect_list:
                self._effect = effect
```

TheJulianJES commented on zha/application/platforms/light/__init__.py:
```diff
@@ -1062,7 +1063,7 @@ def restore_external_state_attributes(
         hs_color: tuple[float, float] | None,
         color_mode: ColorMode | None,
-        effect: str | None,
+        effect: str,
```
`restore_external_state_attributes` can still get `None` from HA, right? E.g. if the light is off or if the `core.restore_state` JSON file is deleted from HA(?)

I think we should revert this change and make sure that `effect is not None` before restoring it below (currently line 1087).

---

"Why should we not restore `None` effect? anymore"
Well, it'll now behave similar to all other state attributes. `None` means "no data" instead of "no effect" now. So, we should follow the same behavior as we do for the other state attributes.
For those, we do not restore `None` because some state attributes may already be populated from the zigpy cache, though it may be slightly outdated for non-attribute reporting lights.

TheJulianJES commented on zha/application/platforms/light/__init__.py:
```diff
@@ -1397,7 +1398,7 @@ def restore_external_state_attributes(
         hs_color: tuple[float, float] | None,
         color_mode: ColorMode | None,
-        effect: str | None,
+        effect: str,
```
Same here, this should be reverted IMO.
```suggestion
        effect: str | None,
```

TheJulianJES committed:
Adjust typing for `effect` in `restore_external_state_attributes`

TheJulianJES committed:
Fix effect behavior in `_assume_group_state`

TheJulianJES committed:
Fix effect behavior in `restore_external_state_attributes`

TheJulianJES commented:
I've rebased the PR to fix the conflicts and addressed the comments I just made above. There's some more context in those comments on why the changes are needed IMO.

TheJulianJES committed:
Adjust `None` state restoration tests for new effect behavior

TheJulianJES commented on tests/test_light.py:
```diff
@@ -2084,12 +2084,12 @@ async def test_light_state_restoration(
         xy_color=None,
         hs_color=None,
         color_mode=None,
-        effect=None,  # Effect is the only `None` value actually restored
```
The workaround of restoring `effect` even if `None` is no longer needed/present, so the tests are adjusted.

TheJulianJES commented on zha/application/platforms/light/__init__.py:
```diff
@@ -710,7 +711,7 @@ def __init__(
             self._min_mireds: int = self._color_cluster_handler.min_mireds
             self._max_mireds: int = self._color_cluster_handler.max_mireds
-        effect_list = []
+        effect_list = [EFFECT_OFF]
```
I believe changing this here might cause the effect option to always appear in HA now, even if the light does not actually support effects (other than off). (Haven't tested though.)

If the light does not support any effects, we do not want to expose `EFFECT_OFF`, correct? (Or do we? Does it matter? Does HA remove not show a single "off" effect in that case maybe?)
We should have a look at this before merging.

Should be addressed with https://github.com/zigpy/zha/pull/99/commits/aa870ef62f908aef14cb1081dbc76fa71a361cd8 (again, not tested, EDIT: reverted for now, see below. Might not matter.)

TheJulianJES replied:
Should be addressed with https://github.com/zigpy/zha/pull/99/commits/aa870ef62f908aef14cb1081dbc76fa71a361cd8
(again, not tested)

TheJulianJES replied:
Ok, I'm not actually sure that is needed, as `_supported_features` will not contain `EFFECT` already.
We may be able to revert that commit. It would also still return `EFFECT_OFF` in `effect`, even though that's not in the supported effects lists with that commit.
Without checking, it's not clear to me what behavior HA expects/wants for the `effect` and `effect_list` property when the "effect" feature isn't supported.

Looking at [HA light code](https://github.com/home-assistant/core/blob/dev/homeassistant/components/light/__init__.py), we may be able to safely revert https://github.com/zigpy/zha/commit/aa870ef62f908aef14cb1081dbc76fa71a361cd8 (?) Both options should really be tested..
EDIT: Reverted for now.

TheJulianJES replied:
Ok, it works fine with that commit reverted.

TheJulianJES committed:
Move out `self._effect_list` into `effect_list`

TheJulianJES committed:
Only add `EFFECT_OFF` if there are other effects

TheJulianJES committed:
Revert "Only add `EFFECT_OFF` if there are other effects"

This reverts commit aa870ef62f908aef14cb1081dbc76fa71a361cd8.

TheJulianJES committed:
Always set effect on members

Changing any other attributes always turns off the effect at the moment, so we need to pass that to the members.

TheJulianJES committed:
Fix group light test expecting `None` `effect_list`

TheJulianJES commented:
I think this PR should be ready now. The missing coverage is on existing lines, something to improve in the future.

---

Also discovered an existing (long-standing) issue when turning off effects via changing the color, but it's unrelated to this PR. Filed https://github.com/zigpy/zha/issues/210. Also something to be fixed in the future. The effect code on/off should be cleaned up a bit with it. Looks like nobody reported this before.
